File: scripts/algorithms/C/Count Primes/Count Primes.js
// Runtime: 439 ms (Top 90.04%) | Memory: 82.1 MB (Top 91.87%)
function makeSieve(n) {
    let arr = new Array(n+1)
    arr[0] = false;
    arr[1] = false;
    arr.fill(true,2,arr.length);

    for( let i = 2; i*i<n; i++) {
        if(arr[i] === true) {
            for( let j = i*i ; j<=n ;j+=i){
                arr[j] = false;
            }
        }
    }

    let count = 0;

    for(let i = 0; i<n;i++) {
        if(arr[i] == true) {
            count++;
        }
    }
    return count;
}

var countPrimes = function(n) {
    let numberOfPrimes = makeSieve(n)
    return numberOfPrimes;
};

File: scripts/algorithms/C/Count Primes/Count Primes.cpp
// Runtime: 258 ms (Top 70.76%) | Memory: 10.2 MB (Top 70.25%)
class Solution {
public:
    int countPrimes(int n) {
        if(n == 0 || n == 1) return 0;
        vector<bool>prime(n,true);
        for(int i = 2; i*i < n; i++){
            if(prime[i]){
                for(int j = i*i; j < n; j += i){
                    prime[j] = false;
                }
            }
        }
        int cnt = 0;
        for(int i = 2; i < n; i++){
            if(prime[i]) cnt++;
        }
        return cnt;
    }
};